dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatcher.java

```diff
@@ -0,0 +1,125 @@
+import org.apache.dolphinscheduler.server.master.runner.queue.PriorityDelayQueue;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * WorkerGroupTaskDispatcher is responsible for dispatching tasks from the task queue.
+ * The main responsibilities include:
+ * 1. Continuously fetching tasks from the {@link PriorityDelayQueue} for dispatch.
+ * 2. Re-queuing tasks that fail to dispatch according to retry logic.
+ * 3. Ensuring thread safety and correct state transitions during task processing.
+ */
+@Slf4j
+public class WorkerGroupTaskDispatcher extends BaseDaemonThread implements AutoCloseable {
+
+    private final ITaskExecutorClient taskExecutorClient;
+
+    // todo: The current queue is flawed. When a high-priority task fails,
+    // it will be delayed and will not return to the first or second position.
+    // Tasks with the same priority will preempt its position.
+    // If it needs to be placed at the front of the queue, the queue needs to be re-implemented.
+    private final PriorityDelayQueue<PriorityAndDelayBasedTaskEntry> workerGroupQueue;
+
+    private final AtomicBoolean RUNNING_FLAG = new AtomicBoolean(false);
+
+    @Getter
+    private DispatchWorkerStatus status;
+
+    public WorkerGroupTaskDispatcher(String workerGroupName, ITaskExecutorClient taskExecutorClient) {
+        super("WorkerGroupTaskDispatcher-" + workerGroupName);
+        this.taskExecutorClient = taskExecutorClient;
+        this.workerGroupQueue = new PriorityDelayQueue<>();
+        status = DispatchWorkerStatus.DEFAULT;
+    }
+
+    public void add(ITaskExecutionRunnable taskExecutionRunnable, long delayTimeMills) {
+        workerGroupQueue.add(new PriorityAndDelayBasedTaskEntry(delayTimeMills, taskExecutionRunnable));
+    }
+
+    public int size() {
+        return workerGroupQueue.size();
+    }
+
+    @Override
+    public synchronized void start() {
+        if (!RUNNING_FLAG.compareAndSet(false, true)) {
+            log.error("The {} already started, will not start again", this.getName());
+            return;
+        }
+        log.info("{} starting...", this.getName());
+        super.start();
+        log.info("{} started...", this.getName());
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (workerGroupQueue.size() == 0) {
+            status = DispatchWorkerStatus.DELETE_SUCCESS;
+            if (RUNNING_FLAG.compareAndSet(true, false)) {
+                log.info("{} stopping...", this.getName());
+                log.info("{} stopped...", this.getName());
+            }
+        } else {
+            log.warn("The {} queue is not empty, will not stop", this.getName());
+            status = DispatchWorkerStatus.DELETING;
+        }
+    }
+
+    @Override
+    public void run() {
+        while (RUNNING_FLAG.get()) {
+            this.dispatch();
+        }
+    }
+
+    public void dispatch() {
```

Comment:
```suggestion
    private void dispatch() {
```

Reply:
This method should not exposed to outside.

Reply:
thanks，i will fix this